### DISCUSSION TITLE
Quantization of pretrained models using onnxruntime

### DISCUSSION THREAD

**Discussion by HemaSowjanyaMamidi** [microsoft/onnxruntime#6387] [Other Q&A]
Hi All,
I took a pretrained model in keras and converted to onnx model and later I wanted to check in terms of quantized models so I tried the following ways.

1. I used onnxruntime-gpu for getting the quantized models but got an error while importing quantize_dynamic method.
2. Later I tried with CPU version of onnxruntime and successfully converted the models where I can observe the change in the size which is reduced by four fold times the original onnx model.
3. So using the converted model I tried to inference on onnxruntime-gpu and found out that the model runs on CPU but not on GPU.

Does the onnxruntime-gpu support the quantized models?
Are the all operators being quantized properly?

**Comment by yufenglee** [comment]
@HemaSowjanyaMamidi, ORT TensorRT EP can run the static quantization, but official onnxruntime-gpu doesn't. Please refer to this script for an e2e example with TRT ep. BTW, the quantization support with TRT EP is still in progress and you may not see performance gain with it:  https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/python/tools/quantization/E2E_example_model/e2e_tensorrt_resnet_example.py.

**Reply by HemaSowjanyaMamidi** [reply]
![results1](https://user-images.githubusercontent.com/47590467/105319360-c45f8000-5bea-11eb-8b63-281869078dca.PNG)
@yufenglee Thanks for the reply. But the quantized models are not even performing well on ort cpu version as well(results are provided in the image). So what can be the reason behind that? Is it because all the operators in the model are not being quantized or should they still need to be optimized even after being quantized

**Reply by yufenglee** [reply]
@HemaSowjanyaMamidi , for CNN model, it is expected to use static quantization as the E2E example here: https://github.com/microsoft/onnxruntime/tree/master/onnxruntime/python/tools/quantization/E2E_example_model/image_classification. 

And could you please share QAT model? It is expected to get similar perf as static quantization.

**Reply by HemaSowjanyaMamidi** [reply]
@yufenglee 
https://github.com/HemaSowjanyaMamidi/onnx_model is the link for qat unoptimized model. I will check the static quantization for cnn.

**Reply by HemaSowjanyaMamidi** [reply]
@yufenglee where can we see the kernel code for quantized operators?

**Reply by yufenglee** [reply]
The customized quantized operator list are here: https://github.com/microsoft/onnxruntime/blob/87ec1f6208f2bb90b079baf444ad942aecfb56ee/onnxruntime/contrib_ops/cpu/cpu_contrib_kernels.cc#L38
https://github.com/microsoft/onnxruntime/blob/87ec1f6208f2bb90b079baf444ad942aecfb56ee/onnxruntime/core/providers/cpu/cpu_execution_provider.cc#L304

**Reply by HemaSowjanyaMamidi** [reply]
@yufenglee Thanks. I tried the static quantized you asked me to. But the results were not great(Below image shows the results i got). So can i know the exact reason behind that one. 
![static_quant](https://user-images.githubusercontent.com/47590467/105485861-89854700-5cd3-11eb-89c3-5535b0282be4.PNG)

**Reply by yufenglee** [reply]
@HemaSowjanyaMamidi , could you please share me your cpu information and your quantized model?

**Reply by HemaSowjanyaMamidi** [reply]
@yufenglee,
**My CPU information:**
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              2
On-line CPU(s) list: 0,1
Thread(s) per core:  2
Core(s) per socket:  1
Socket(s):           1
NUMA node(s):        1
Vendor ID:           GenuineIntel
CPU family:          6
Model:               63
Model name:          Intel(R) Xeon(R) CPU @ 2.30GHz
Stepping:            0
CPU MHz:             2299.998
BogoMIPS:            4599.99
Hypervisor vendor:   KVM
Virtualization type: full
L1d cache:           32K
L1i cache:           32K
L2 cache:            256K
L3 cache:            46080K
NUMA node0 CPU(s):   0,1

**Quantized model link:**
https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/python/tools/quantization/E2E_example_model/image_classification/mobilenetv2-7.onnx
This is the model I was using.

**Reply by HemaSowjanyaMamidi** [reply]
@yufenglee, can you please let me know if there are any specific hardware required for the quantized models to run? and I observed that the models converted from other frameworks to onnx after being quantized take longer time for inference than the onnx model obtained after conversion. Are my observations correct in this regard..? Or should we change something after the models are converted from source framework to onnx?

**Reply by yufenglee** [reply]
@HemaSowjanyaMamidi, quantized models are expected to get better performance on CPU with [VNNI extension](https://en.wikichip.org/wiki/x86/avx512_vnni). On CPU without VNNI, it depends on the model. If the model is IO bounding, it may get better performance; if compute bounding, it can not in general.

**Reply by HemaSowjanyaMamidi** [reply]
Thanks @yufenglee, how can we know whether a model is compute bounding or IO bounding.
